modules/lsm/lsm_local.c

```diff
@@ -0,0 +1,255 @@
+ *
+ */
+
+/* Note: This is inspired by modules/dummy/dummylinuxdrive.c  */
+
+#include "config.h"
+
+#include <sys/types.h>
+
+#include <pwd.h>
+#include <grp.h>
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <glib.h>
+#include <libstoragemgmt/libstoragemgmt.h>
+
+#include <src/udiskslogging.h>
+#include <src/udiskslinuxdriveobject.h>
+#include <src/udisksdaemon.h>
+#include <src/udisksdaemonutil.h>
+#include <src/udisksbasejob.h>
+#include <src/udiskssimplejob.h>
+#include <src/udisksthreadedjob.h>
+#include <src/udiskslinuxdevice.h>
+#include <modules/udisksmoduleobject.h>
+
+#include "lsm_types.h"
+
+typedef struct _UDisksLinuxDriveLsmLocalClass
+  UDisksLinuxDriveLsmLocalClass;
+
+struct _UDisksLinuxDriveLsmLocal
+{
+  UDisksDriveLsmLocalSkeleton parent_instance;
+  UDisksLinuxDriveObject *ud_lx_drv_obj;
+};
+
+struct _UDisksLinuxDriveLsmLocalClass
+{
+  UDisksDriveLsmLocalSkeletonClass parent_class;
+};
+
+static void
+udisks_linux_drive_lsm_local_iface_init
+  (UDisksDriveLsmLocalIface *iface);
+
+G_DEFINE_TYPE_WITH_CODE
+  (UDisksLinuxDriveLsmLocal, udisks_linux_drive_lsm_local,
+   UDISKS_TYPE_DRIVE_LSM_LOCAL_SKELETON,
+   G_IMPLEMENT_INTERFACE (UDISKS_TYPE_DRIVE_LSM_LOCAL,
+                          udisks_linux_drive_lsm_local_iface_init));
+
+static const gchar *
+get_blk_path (UDisksDriveLsmLocal *ud_drv_lsm_local,
+               GDBusMethodInvocation *invocation);
+
+static gboolean
+handle_turn_ident_ledon (UDisksDriveLsmLocal *ud_drv_lsm_local,
+                         GDBusMethodInvocation *invocation);
+
+static gboolean
+handle_turn_ident_ledoff (UDisksDriveLsmLocal *ud_drv_lsm_local,
+                          GDBusMethodInvocation *invocation);
+
+static gboolean
+handle_turn_fault_ledon (UDisksDriveLsmLocal *ud_drv_lsm_local,
+                         GDBusMethodInvocation *invocation);
+
+static gboolean
+handle_turn_fault_ledoff (UDisksDriveLsmLocal *ud_drv_lsm_local,
+                          GDBusMethodInvocation *invocation);
+
+static const gchar *
+get_blk_path (UDisksDriveLsmLocal *ud_drv_lsm_local,
+              GDBusMethodInvocation *invocation)
+{
+  const gchar *blk_path = NULL;
+  UDisksLinuxDriveLsmLocal *ud_lx_drv_lsm_local = NULL;
+  UDisksLinuxDriveObject *ud_lx_drv_obj = NULL;
+  UDisksLinuxBlockObject *ud_lx_blk_obj = NULL;
+  UDisksBlock *ud_blk = NULL;
+
+  ud_lx_drv_lsm_local = UDISKS_LINUX_DRIVE_LSM_LOCAL (ud_drv_lsm_local);
+
+  ud_lx_drv_obj = ud_lx_drv_lsm_local->ud_lx_drv_obj;
+
+  ud_lx_blk_obj = udisks_linux_drive_object_get_block (ud_lx_drv_obj, FALSE);
+  if (ud_lx_blk_obj == NULL)
+    {
+      g_dbus_method_invocation_return_error
+        (invocation, UDISKS_ERROR, UDISKS_ERROR_FAILED,
+         "Unable to find block device for drive");
+      goto out;
+    }
+
+  ud_blk = udisks_object_peek_block (UDISKS_OBJECT (ud_lx_blk_obj));
+  blk_path = g_strdup (udisks_block_get_device (ud_blk));
```

Comment:
Please check the of the ``peek_block()`` [1] and ``get_device()`` [2] functions above. I believe they should be replaced by their thread-safe/ref-counting equivalents.

[1] http://storaged.org/doc/udisks2-api/UDisksObject.html#udisks-object-peek-block
[2] http://storaged.org/doc/udisks2-api/UDisksBlock.html#udisks-block-get-device